pkg/engine/lifecycletest/step_generator_test.go

```diff
@@ -92,3 +93,107 @@ func TestDuplicateAlias(t *testing.T) {
+	t.Parallel()
+
+	program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		parentUrn, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			DeletedWith: "expect-this",
+		})
+		assert.NoError(t, err)
+
+		_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resB", true, deploytest.ResourceOptions{
+			Parent: parentUrn,
+		})
+		assert.NoError(t, err)
+
+		return nil
+	})
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				DiffF: func(
+					urn resource.URN, id resource.ID, olds, news resource.PropertyMap, ignoreChanges []string,
+				) (plugin.DiffResult, error) {
+					return plugin.DiffResult{}, nil
+				},
+			}, nil
+		}),
+	}
+
+	host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+
+	p := &TestPlan{
+		Options: UpdateOptions{Host: host},
+	}
+
+	project := p.GetProject()
+	snap, res := TestOp(Update).Run(project, p.GetTarget(t, nil), p.Options, false, p.BackendClient, nil)
+	for _, res := range snap.Resources[1:] {
+		assert.Equal(t, resource.URN("expect-this"), res.DeletedWith)
+	}
+	assert.Nil(t, res)
+}
+
+// TestDeletedWithOptionInheritanceMLC checks that an MLC's DeletedWith option is propagated to resources that
+// set an MLC as its parent. MLC's are remote and at the time of writing their RegisterResource call asks the
+// resource monitor to ask the constructor to call the necessary RegisterResource calls on the program's behalf.
+func TestDeletedWithOptionInheritanceMLC(t *testing.T) {
+	t.Parallel()
+
+	program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		parentUrn, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			Remote:      true,
+			DeletedWith: "expect-this",
+		})
+		assert.NoError(t, err)
+
+		_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resB", true, deploytest.ResourceOptions{
+			Parent: parentUrn,
+		})
+		assert.NoError(t, err)
+
+		return nil
+	})
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				DiffF: func(
+					urn resource.URN, id resource.ID, olds, news resource.PropertyMap, ignoreChanges []string,
+				) (plugin.DiffResult, error) {
+					return plugin.DiffResult{}, nil
+				},
+				ConstructF: func(monitor *deploytest.ResourceMonitor, typ, name string,
+					parent resource.URN, inputs resource.PropertyMap, options plugin.ConstructOptions,
+				) (plugin.ConstructResult, error) {
+					urn, _, _, err := monitor.RegisterResource(tokens.Type(typ), name, false, deploytest.ResourceOptions{
+						DeletedWith: options.DeletedWith,
+					})
+					assert.NoError(t, err)
+					assert.Equal(t, "resA", name)
```

Comment:
Consider moving this to the beginning of the `ConstructF` function, and changing it to a `require` rather than `assert` so it immediately fails the test. Also, might as well assert that the `typ` is the expected type (if you change it based on my other comment above, `pkgA:m:typComponent`:

```go
require.Equal(t, "pkgA:m:typComponent", typ)
require.Equal(t, "resA", name)
```